Framework/DataHandling/src/MaskDetectors.cpp

```diff
@@ -260,6 +217,82 @@ void MaskDetectors::exec() {
     setProperty("Workspace", ws);
   }
 }
+
+/** Handle masking a workspace using a MaskWorkspace.
+ *
+ * This will choose the strategy to mask the input workspace with based on
+ * information about the mask workspace.
+ *
+ * @param maskWs :: pointer to the mask workspace to use.
+ * @param WS :: pointer to the workspace to be masked.
+ * @param detectorList :: list of detectors to be masked.
+ * @param indexList :: list of workspace indicies to be masked.
+ * @param rangeInfo :: information about the spectrum range to use when masking
+ */
+void MaskDetectors::handleMaskByMaskWorkspace(
+    const MaskWorkspace_const_sptr maskWs,
+    const API::MatrixWorkspace_const_sptr WS,
+    std::vector<detid_t> &detectorList, std::vector<size_t> &indexList,
+    const RangeInfo &rangeInfo) {
+
+  if (maskWs->getInstrument()->getDetectorIDs().size() !=
+      WS->getInstrument()->getDetectorIDs().size()) {
+    throw std::runtime_error("Instrument's detector numbers mismatch "
+                             "between input Workspace and MaskWorkspace");
+  }
+
+  g_log.debug() << "Extracting mask from MaskWorkspace (" << maskWs->getName()
+                << ")\n";
+  bool forceDetIDs = getProperty("ForceInstrumentMasking");
+  if (maskWs->getNumberHistograms() != WS->getNumberHistograms() ||
+      forceDetIDs) {
+    extractMaskedWSDetIDs(detectorList, maskWs);
+  } else {
+    appendToIndexListFromMaskWS(indexList, maskWs, rangeInfo);
+  }
+}
+
+/** Handle masking a workspace using a MatrixWorkspace.
+ *
+ * This will choose the strategy to mask the input workspace with based on
+ * information about the matrix workspace passed as a mask workspace.
+ *
+ * @param maskWs :: pointer to the workspace to use as a mask.
+ * @param WS :: pointer to the workspace to be masked.
+ * @param detectorList :: list of detectors to be masked.
+ * @param indexList :: list of workspace indicies to be masked.
+ * @param rangeInfo :: information about the spectrum range to use when masking
+ */
+void MaskDetectors::handleMaskByMatrixWorkspace(
+    const API::MatrixWorkspace_const_sptr maskWs,
+    const API::MatrixWorkspace_const_sptr WS,
+    std::vector<detid_t> &detectorList, std::vector<size_t> &indexList,
+    const RangeInfo &rangeInfo) {
+
+  const auto nHist = maskWs->getNumberHistograms();
+  auto instrument = WS->getInstrument();
+  auto maskInstrument = maskWs->getInstrument();
+
+  // Check the provided workspace has the same number of spectra as the
+  // input, if so assume index list
+  if (nHist == WS->getNumberHistograms()) {
+    appendToIndexListFromWS(indexList, maskWs, rangeInfo);
+    // Check they both have instrument and then use detector based masking
+  } else if (instrument && maskInstrument) {
```

Comment:
This should check the instrument->name, not pointer address, but it also should only log as warning if they do not match, still go ahead and apply the masking.